test/hts-precompile/crypto-allowance/cryptoAllowance.js

```diff
@@ -0,0 +1,237 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+const utils = require('../utils');
+const Utils = require('../utils');
+const { expect } = require('chai');
+const { ethers } = require('hardhat');
+const Constants = require('../../constants');
+const {
+  pollForNewSignerBalanceUsingProvider,
+} = require('../../../utils/helpers');
+
+describe('@CryptoAllowance Test Suite', () => {
+  let walletA, walletB, walletC, cryptoAllowanceContract;
+  const amount = 3000;
+
+  before(async () => {
+    [walletA, walletB, walletC, receiver] = await ethers.getSigners();
+
+    const CryptoAllowanceFactory = await ethers.getContractFactory(
+      Constants.Contract.CryptoAllowance
+    );
+
+    cryptoAllowanceContract = await CryptoAllowanceFactory.deploy();
+    await cryptoAllowanceContract.waitForDeployment();
+  });
+
+  it('Should execute hbarApprovePublic and return success response code', async () => {
+    const tx = await cryptoAllowanceContract.hbarApprovePublic(
+      cryptoAllowanceContract.target,
+      walletA.address,
+      amount,
+      Constants.GAS_LIMIT_1_000_000
+    );
+    const receipt = await tx.wait();
+    const responseCode = receipt.logs.find(
+      (l) => l.fragment.name === 'ResponseCode'
+    );
+    expect(responseCode.args).to.deep.eq([22n]);
+  });
+
+  it('Should execute hbarAllowancePublic and return an event with the allowance information', async () => {
+    const approveTx = await cryptoAllowanceContract.hbarApprovePublic(
+      cryptoAllowanceContract.target,
+      walletB.address,
+      amount,
+      Constants.GAS_LIMIT_1_000_000
+    );
+    await approveTx.wait();
+
+    const allowanceTx = await cryptoAllowanceContract.hbarAllowancePublic(
+      cryptoAllowanceContract.target,
+      walletB.address,
+      Constants.GAS_LIMIT_1_000_000
+    );
+
+    const receipt = await allowanceTx.wait();
+    const responseCode = receipt.logs.find(
+      (l) => l.fragment.name === 'ResponseCode'
+    );
+    const logs = receipt.logs.find((l) => l.fragment.name === 'HbarAllowance');
+
+    expect(responseCode.args).to.deep.eq([22n]);
+    expect(logs.args[0]).to.eq(cryptoAllowanceContract.target);
+    expect(logs.args[1]).to.eq(walletB.address);
+    expect(logs.args[2]).to.eq(amount);
+  });
+
+  it('Should allow an approval on behalf of hbar owner WITH its signature', async () => {
+    // update accountKeys
+    const ecdsaPrivateKeys = await Utils.getHardhatSignersPrivateKeys(false);
+    await utils.updateAccountKeysViaHapi(
+      [cryptoAllowanceContract.target],
+      [ecdsaPrivateKeys[0]] // walletA's key
```

Comment:
If I understand this method correctly, are we updating Wallet A key with the `cryptoAllowanceContract.target` address? 

why do we need to do this, I thought that the purpose of CryptoAllowance was to get rid of the need to update the keys as per the issue description here: https://github.com/hashgraph/hedera-smart-contracts/issues/797

Reply:
> If I understand this method correctly, are we updating Wallet A key with the cryptoAllowanceContract.target address?

Yes, you're correct! The `updateAccountKeysViaHapi` indeed adds the `cryptoAllowanceContract.target` to the Wallet A's key. The reason why we need to include `cryptoAllowanceContract` contractId to Wallet A's key is because CryptoAllowance is a wrapper contract exposing the HederaAccountService (HAS) precompiled system contract. Therefore, this wrapper contract, not the `owner address`, will eventually be the top level caller initializing the cryptoTransfer against the HAS precompile contract. Hence, it needs the right key to sign the transaction. 

> I thought that the purpose of CryptoAllowance was to get rid of the need to update the keys as per the issue description here: https://github.com/hashgraph/hedera-smart-contracts/issues/797

Yes and no. `CryptoAllowance` is the wrapper contract so its purpose is to grant hbar allowance from one account to another account but still need the right key to sign based on the security model. The part `get rid of the need to update the keys` is actually done by [IHRC632](https://github.com/hashgraph/hedera-smart-contracts/pull/800/files#diff-b1714211ffddee036d407bd59b55a27f0e5b415ca95e5bab6e69bbc2dd474732), which was also mentioned in HIP-906. IHRC632 allows an EOA to act as an SC with IHRC632 interface, `walletAIHrc632 = new Contract(walletA.address, IHRC632, walletA);`. This will make it possible for the EOA to directly make calls to the HAS precompile contracts, hence making the EOA top level caller of the transaction without any help from third party like another SC account. That's when Hip-906 kicks in and allow hbar owner, top level caller, grant an hbar allowance to a spender.

Reply:
Thanks for the explanation, it was hard to review the PR due to all that context that needed to be digested and I think I got a grasp of it and that is why I bring those questions up. It makes sense now. Thanks 👍